compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala

```diff
@@ -185,15 +187,28 @@ object ProtoTypes {
     /** A map recording the typer states in which arguments stored in myTypedArg were typed */
     private var evalState: SimpleMap[untpd.Tree, TyperState] = SimpleMap.Empty
 
-    def isMatchedBy(tp: Type)(implicit ctx: Context) =
+    private[this] var thisCallArgCtxUsed: Boolean = false
+
+    /** Evalute `op` with the proper context to type the arguments. */
+    protected def withArgCtx[T](op: Context => T)(implicit ctx: Context) = {
+      if (isSelfConstrCall && !thisCallArgCtxUsed) {
+        // Context#thisCallArgContext is not idempotent so we must only use it
```

Comment:
Can we use `Mode.InSuperCall` as an indicator whether we are already in the right context, instead of defining a special field for this?